src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+                    solutionAfterFirstMerge,
+                    _state.DocumentIds, 
+                    CreateAllContainingNamespaces(OldNamespaceParts), 
+                    cancellationToken).ConfigureAwait(false);
+
+                solutionAfterImportsRemoved = await RemoveUnnecessaryImportsAsync(
+                    solutionAfterImportsRemoved, 
+                    referenceDocuments.ToImmutableArray(), 
+                    ImmutableArray.Create(_state.DeclaredNamespace, _state.TargetNamespace), 
+                    cancellationToken).ConfigureAwait(false);
+
+                var solutionAfterSecondMerge = await MergeDiffAsync(solutionAfterFirstMerge, solutionAfterImportsRemoved, cancellationToken).ConfigureAwait(false);
+
+                referenceDocuments.Free();
+                return ImmutableArray.Create<CodeActionOperation>(new ApplyChangesOperation(solutionAfterSecondMerge));
+            }
+
+            private ImmutableArray<string> CreateAllContainingNamespaces(ImmutableArray<string> parts)
+            {
+                var builder = ArrayBuilder<string>.GetInstance();
+                for (var i = 1; i <= OldNamespaceParts.Length; ++i)
+                {
+                    builder.Add(string.Join(".", OldNamespaceParts.Take(i)));
+                }
+
+                return builder.ToImmutableAndFree();
+            }
+
+            private ImmutableArray<SyntaxNode> CreateImports(Document document, ImmutableArray<string> names, bool withFormatterAnnotation)
+            {
+                var generator = SyntaxGenerator.GetGenerator(document);
+                var builder = ArrayBuilder<SyntaxNode>.GetInstance(names.Length);
+                for (var i = 0; i < names.Length; ++i)
+                {
+                    builder.Add(CreateImport(generator, names[i], withFormatterAnnotation));
+                }
+
+                return builder.ToImmutableAndFree();
+            }
+
+            private SyntaxNode CreateImport(SyntaxGenerator syntaxGenerator, string name, bool withFormatterAnnotation)
+            {
+                var import = syntaxGenerator.NamespaceImportDeclaration(name);
+                if (withFormatterAnnotation)
+                {
+                    import = import.WithAdditionalAnnotations(Formatter.Annotation);
+                }
+                return import;
+            }
+
+            private async Task<(Solution, ImmutableArray<DocumentId>)> ChangeNamespaceToMatchFoldersAsync(Solution solution, DocumentId id, CancellationToken cancellationToken)
+            {
+                var document = solution.GetDocument(id);
+                var targetNamespace = _state.TargetNamespace;
+
+                var declarationRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var container = declarationRoot.DescendantNodes().FirstOrDefault(node => node is TNamespaceDeclarationSyntax) ?? declarationRoot;
```

Comment:
anotehr case where i'm guessing things would be cleaner if the compilationunit/namespacedecl were just computed and passed along.

i think i've seen this DescendentNodes code in about 3 places now. That's pretty brittle as, effectively, all these different functions (even in different classes) have to agree on this behavior.  it's nicer to just have it in one location, that htey can all depend on.

Reply:
Sure, but I'd rather do it as part of #30896, more changes will be needed in language service and code action.